samples/backtest_grid.py

```diff
@@ -0,0 +1,330 @@
+import dataclasses
+import datetime
+import logging
+from decimal import Decimal
+from typing import Dict, Optional, List
+
+import basana as bs
+import basana.backtesting.exchange as backtesting_exchange
+from basana import OrderOperation, Pair
+from basana.backtesting import charts, lending
+from basana.core.event_sources.trading_signal import BaseTradingSignal
+from basana.core.logs import StructuredMessage
+from basana.external.binance import csv
+from samples.configs.grid_config import Config
+from samples.strategies import bbands
+
+
+@dataclasses.dataclass
+class PositionInfo:
+    """
+    PositionInfo class
+    """
+    pair: Pair
+    signal_position: bs.Position
+
+
+class Point:
+    def __init__(self, pair: Pair):
+        """
+        Init Point instance
+
+        @param pair:
+        """
+        self._value = 0.0001 if pair.quote_symbol.lower() != "JPY".lower() else 0.01
+
+    @property
+    def value(self):
+        return self._value
+
+
+class GridPositionManager:
+    """
+    GridPositionManager class
+    """
+
+    # Responsible for managing orders and tracking positions in response to trading signals.
+    def __init__(
+            self, exchange: backtesting_exchange.Exchange, position_amount: Decimal, grid_size: int, grid_step: int,
+            hedging: bool, martingale_sizes: bool, borrowing_disabled: bool = False
+    ):
+        """
+        Init GridPositionManager instance
+
+        Manage a grid system
+
+        @param exchange: Exchange instance
+        @param position_amount: Amount of base value to trade
+        @param grid_size: Grid number of lines above and below the start price
+        @param grid_step: Space in Pips between grid lines
+        @param hedging: True to take hedging orders
+        @param martingale_sizes: True to increase position sizes in grid
+        @param borrowing_disabled: Disable loans on platform (ie: do not use short trades), default= False
+        """
+        assert position_amount > 0
+
+        # configuration
+        self._exchange = exchange
+        self._position_amount = position_amount
+        self._grid_size = grid_size
+        self._grid_step = grid_step
+        self._hedging = hedging
+        self._martingale_sizes = martingale_sizes
+        self._borrowing_disabled = borrowing_disabled
+
+        # storage lists
+        self._positions: Dict[bs.Pair, PositionInfo] = {}
+        self._buy_orders_ids: List[str] = []
+        self._sell_orders_ids: List[str] = []
+
+    async def on_trading_signal(self, trading_signal: BaseTradingSignal):
+        """
+        Handle trading signals
+
+        @param trading_signal: TradingSignal instance
+        """
+        pairs = list(trading_signal.get_pairs())
+        logging.info(StructuredMessage("Trading signal", pairs=pairs))
+
+        try:
+            coros = []
+            for pair, target_position in pairs:
+                if self._borrowing_disabled and target_position == bs.Position.SHORT:
```

Comment:
I don't think that you wan't to map SHORT to NEUTRAL for this kind of strategy. You'll end up placing a grid of orders when the market is trending down.